tests/test_ext_viewcode.py

```diff
@@ -60,3 +62,21 @@ def test_linkcode(app, status, warning):
     assert 'http://foobar/js/' in stuff
     assert 'http://foobar/c/' in stuff
     assert 'http://foobar/cpp/' in stuff
+
+
+@pytest.mark.sphinx(testroot='ext-viewcode', tags=['test_source_files'])
+def test_local_source_files(app, status, warning):
+    # Let autodoc document the module,
+    # but unload it so that viewcode has to use the source files.
+    def unload_spam(app, doctree):
+        sys.path.pop(0)
+        for module in list(sys.modules):
+            if module.startswith('spam.') or module == 'spam':
+                del sys.modules[module]
+
+    listeners = app.events.listeners['doctree-read']
+    viewcode_read = listeners.popitem()[1]
+    app.connect('doctree-read', unload_spam)
+    app.connect('doctree-read', viewcode_read)
```

Comment:
I feel this code is very hacky. I can't understand what this code doing intuitively. How about adding new test project? Then we don't hack `sys.path` and event listeners.